observ/store.py

```diff
@@ -0,0 +1,135 @@
+import copy
+from functools import partial, wraps
+from inspect import signature
+from typing import Callable, TypeVar
+
+from .dep import Dep
+from .observables import (
+    reactive,
+    readonly,
+    shallow_reactive,
+    to_raw,
+)
+from .watcher import Watcher
+
+
+T = TypeVar("T", bound=Callable)
+
+
+registry = {}
+
+
+def mutation(fn: T) -> T:
+    registry[fn] = "mutation"
+
+    @wraps(fn)
+    def inner(self, *args, **kwargs):
+        return fn(self, *args, **kwargs)
+
+    return inner
+
+
+def computed(fn: T) -> T:
+    parameters = signature(fn).parameters
+    if len(parameters) == 1 and "self" in parameters:
+        registry[fn] = "computed"
+
+        @wraps(fn)
+        def inner(self):
+            return fn(self)
+
+        return inner
+    else:
+        watcher = Watcher(fn)
+
+        @wraps(fn)
+        def getter():
+            if watcher.dirty:
+                watcher.evaluate()
+            if Dep.stack:
+                watcher.depend()
+            return watcher.value
+
+        getter.__watcher__ = watcher
+        return getter
+
+
+registered_computed_props = {}
+
+
+class Store:
+    """
+    Store that tracks mutations to state in order to enable undo/redo functionality
+    """
+
+    def __init__(self, state):
+        self._present = reactive(state)
+        # Because we can't work with diff patches (yet) we'll have to copy
+        # over the whole state. When triggering undo/redo, we'll therefore
+        # trigger all the watchers / computed expressions...
+        self._past = shallow_reactive([copy.deepcopy(to_raw(state))])
+        self._future = shallow_reactive([])
+        self.state = readonly(state)
+
+        computed_props = registered_computed_props.setdefault(id(self), {})
+        for method_name in dir(self):
+            method = getattr(self, method_name)
+            if not hasattr(method, "__wrapped__"):
```

Comment:
Why not check if the method is in `registry`? Seems more appropriate!

Reply:
That's way more simple!